kernel/src/fs/notify/inotify.rs

```diff
@@ -0,0 +1,587 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{
+    collections::VecDeque,
+    string::String,
+    sync::{Arc, Weak},
+};
+use core::{
+    any::Any,
+    sync::atomic::{AtomicBool, AtomicU32, AtomicU64, Ordering},
+};
+
+use bitflags::bitflags;
+use hashbrown::HashMap;
+use ostd::{mm::VmWriter, sync::Mutex};
+
+use crate::{
+    current_userspace,
+    events::IoEvents,
+    fs::{
+        file_handle::FileLike,
+        notify::{FsnotifyEvents, FsnotifySubscriber},
+        path::Path,
+        utils::{AccessMode, Inode, InodeMode, IoctlCmd, Metadata, StatusFlags},
+    },
+    prelude::*,
+    process::signal::{PollHandle, Pollable, Pollee},
+    return_errno_with_message,
+};
+
+#[derive(Clone)]
+struct SubscriberEntry {
+    inode: Weak<dyn Inode>,
+    subscriber: Weak<dyn FsnotifySubscriber>,
+}
+
+/// A file-like object that provides inotify functionality.
+///
+/// InotifyFile accepts events from multiple inotify subscribers (watches) on different inodes.
+/// Users should read events from this file to receive notifications about file system changes.
+pub struct InotifyFile {
+    // Lock to serialize subscriber updates and removals.
+    subscriber_lock: Mutex<()>,
+    // A subscriber descriptor allocator.
+    sd_allocator: AtomicU32,
```

Comment:
Do not use invent a new obselete abbreviation, "sd". Linux already chooses the term "watch" and "wd".
  
Do not confuse external and internal concepts. Let's follow Linux's terminology at least at the interface level of `InotifyFile`. Adopt the concepts of "watch" and "wd". Do not expose the internal concepts of subscribers to the interface (`update_subscriber` or `remove_subscriber`). They are implementation details.

In addition, this field is not an allocator by itself.

In summary, a more appropriate name is `next_wd`.